### PR TITLE
Add support for IBM cloud provider

### PR DESCRIPTION
**Pull request by anilkumarnagaraj**
This PR adds support for IBM CLOUD provider. 

**List of supported resources:**

*   `ibm_cos`
    * `ibm_resource_instance`
    * `ibm_cos_bucket`
*   `ibm_iam`
    * `ibm_iam_user_policy`
    * `ibm_iam_access_group`
    * `ibm_iam_access_group_members`
    * `ibm_iam_access_group_policy`
    * `ibm_iam_access_group_dynamic_rule`
*   `ibm_container_vpc_cluster`
    * `ibm_container_vpc_cluster`
    * `ibm_container_vpc_worker_pool`
*   `ibm_database_etcd`
    * `ibm_database`
*   `ibm_database_mongo`
    * `ibm_database`
*   `ibm_database_postgresql`
    * `ibm_database`
*   `ibm_database_rabbitmq`
    * `ibm_database`
*   `ibm_database_redis`
    * `ibm_database`
*   `ibm_is_instance_group`
    * `ibm_is_instance_group`
    * `ibm_is_instance_group_manager`
    * `ibm_is_instance_group_manager_policy`
*   `ibm_is_vpc`
    * `ibm_is_vpc`
    * `ibm_is_vpc_address_prefix`
    * `ibm_is_vpc_route`
*   `ibm_is_subnet`
*   `ibm_is_instance`
*   `ibm_is_security_group`
    * `ibm_is_security_group_rule`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -859,6 +860,77 @@ List of supported AliCloud resources:
 * `vswitch`
   * `alicloud_vswitch`
 
+ ### Use with IBM Cloud
+
+If you want to run Terraformer with the IBM Cloud provider plugin on your system, complete the following steps:
```
Comment (sergeylanzman) [inline README.md:865]: It's same process for all providers, you can delete 1-6.

File: providers/ibm/cos.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	authEndpoint := "https://iam.cloud.ibm.com/identity/token"
+	for _, cs := range cosInstances {
+		g.Resources = append(g.Resources, g.loadCOS(cs.ID, cs.Name))
+		s3Conf := ibmaws.NewConfig().WithCredentials(ibmiam.NewStaticCredentials(ibmaws.NewConfig(), authEndpoint, os.Getenv("IC_API_KEY"), cs.ID)).WithS3ForcePathStyle(true).WithEndpoint("s3.us-south.cloud-object-storage.appdomain.cloud")
+		s3Sess := cossession.Must(cossession.NewSession())
+		s3Client := coss3.New(s3Sess, s3Conf)
+		d, _ := s3Client.ListBucketsExtended(&coss3.ListBucketsExtendedInput{})
+		for _, b := range d.Buckets {
+			var apiType, location string
+			singleSiteLocationRegex, _ := regexp.Compile("^[a-z]{3}[0-9][0-9]-[a-z]{4,8}$")
```
Comment (sergeylanzman) [inline providers/ibm/cos.go:100]: Can you move regex outside loops? for better performance

File: providers/ibm/database_elasticsearch.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+	"os"
+
+	"github.com/GoogleCloudPlatform/terraformer/terraformutils"
+	bluemix "github.com/IBM-Cloud/bluemix-go"
+	"github.com/IBM-Cloud/bluemix-go/api/resource/resourcev1/catalog"
+	"github.com/IBM-Cloud/bluemix-go/api/resource/resourcev2/controllerv2"
+	"github.com/IBM-Cloud/bluemix-go/session"
+)
+
+// Copyright 2019 The Terraformer Authors.
```
Comment (sergeylanzman) [inline providers/ibm/database_elasticsearch.go:13]: License first

File: providers/ibm/database_etcd.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	"os"
+
+	"github.com/GoogleCloudPlatform/terraformer/terraformutils"
+	bluemix "github.com/IBM-Cloud/bluemix-go"
+	"github.com/IBM-Cloud/bluemix-go/api/resource/resourcev1/catalog"
+	"github.com/IBM-Cloud/bluemix-go/api/resource/resourcev2/controllerv2"
+	"github.com/IBM-Cloud/bluemix-go/session"
+)
+
+// Copyright 2019 The Terraformer Authors.
```
Comment (sergeylanzman) [inline providers/ibm/database_etcd.go:13]: License first

File: providers/ibm/database_mongo.go [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+	"os"
+
+	"github.com/GoogleCloudPlatform/terraformer/terraformutils"
+	bluemix "github.com/IBM-Cloud/bluemix-go"
+	"github.com/IBM-Cloud/bluemix-go/api/resource/resourcev1/catalog"
+	"github.com/IBM-Cloud/bluemix-go/api/resource/resourcev2/controllerv2"
+	"github.com/IBM-Cloud/bluemix-go/session"
+)
+
+// Copyright 2019 The Terraformer Authors.
```
Comment (sergeylanzman) [inline providers/ibm/database_mongo.go:13]: License first

File: providers/ibm/database_postgresql.go [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/GoogleCloudPlatform/terraformer/terraformutils"
+	bluemix "github.com/IBM-Cloud/bluemix-go"
+	"github.com/IBM-Cloud/bluemix-go/api/resource/resourcev1/catalog"
+	"github.com/IBM-Cloud/bluemix-go/api/resource/resourcev2/controllerv2"
+	"github.com/IBM-Cloud/bluemix-go/session"
+)
+
+// Copyright 2019 The Terraformer Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (sergeylanzman) [inline providers/ibm/database_postgresql.go:16]: same

File: providers/ibm/database_postgresql.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+// loadPostgresqlDB ...
+func (g DatabasePostgresqlGenerator) loadPostgresqlDB(dbID string, dbName string) terraformutils.Resource {
+	var resources terraformutils.Resource
+	resources = terraformutils.NewSimpleResource(
+		dbID,
+		dbName,
+		"ibm_database",
+		"ibm",
+		[]string{})
+	log.Printf("Resources list %+v", resources)
```
Comment (sergeylanzman) [inline providers/ibm/database_postgresql.go:42]: delete log.Printf

File: providers/ibm/database_rabbitmq.go [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+	"github.com/GoogleCloudPlatform/terraformer/terraformutils"
+	bluemix "github.com/IBM-Cloud/bluemix-go"
+	"github.com/IBM-Cloud/bluemix-go/api/resource/resourcev1/catalog"
+	"github.com/IBM-Cloud/bluemix-go/api/resource/resourcev2/controllerv2"
+	"github.com/IBM-Cloud/bluemix-go/session"
+)
+
+// Copyright 2019 The Terraformer Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (sergeylanzman) [inline providers/ibm/database_rabbitmq.go:15]: same

File: providers/ibm/iam.go [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+		"ibm",
+		[]string{})
+	return resources
+}
+
+func (g IAMGenerator) loadAccessGroupMembers(grpID string) terraformutils.Resource {
+	var resources terraformutils.Resource
+	resources = terraformutils.NewSimpleResource(
+		fmt.Sprintf("%s/%s", grpID, time.Now().UTC().String()),
+		time.Now().UTC().String(),
```
Comment (sergeylanzman) [inline providers/ibm/iam.go:61]: It's generate every time other IDs, Maybe better take hash or combine.

Some people run terraformer many time and check git diff

File: providers/ibm/ibm_is_instance.go [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+		[]string{})
+	return resources
+}
+
+func (g *InstanceGenerator) InitResources() error {
+	var resoureGroup string
+	region := envFallBack([]string{"IC_REGION"}, "us-south")
+	apiKey := os.Getenv("IC_API_KEY")
+	if apiKey == "" {
+		log.Fatal("No API key set")
```
Comment (sergeylanzman) [inline providers/ibm/ibm_is_instance.go:47]: no fatal, please return error

File: providers/ibm/ibm_is_security_group.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+		[]string{})
+	return resources
+}
+
+func (g *SecurityGroupGenerator) InitResources() error {
+	var resoureGroup string
+	region := envFallBack([]string{"IC_REGION"}, "us-south")
+	apiKey := os.Getenv("IC_API_KEY")
+	if apiKey == "" {
+		log.Fatal("No API key set")
```
Comment (sergeylanzman) [inline providers/ibm/ibm_is_security_group.go:59]: return error

File: providers/ibm/ibm_is_subnet.go [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+		[]string{})
+	return resources
+}
+
+func (g *SubnetGenerator) InitResources() error {
+	var resoureGroup string
+	region := envFallBack([]string{"IC_REGION"}, "us-south")
+	apiKey := os.Getenv("IC_API_KEY")
+	if apiKey == "" {
+		log.Fatal("No API key set")
```
Comment (sergeylanzman) [inline providers/ibm/ibm_is_subnet.go:47]: return error

File: providers/ibm/ibm_is_vpc.go [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+		[]string{})
+	return resources
+}
+
+func (g *VPCGenerator) InitResources() error {
+	var resoureGroup string
+	region := envFallBack([]string{"IC_REGION"}, "us-south")
+	apiKey := os.Getenv("IC_API_KEY")
+	if apiKey == "" {
+		log.Fatal("No API key set")
```
Comment (sergeylanzman) [inline providers/ibm/ibm_is_vpc.go:69]: return error

File: providers/ibm/ibm_provider.go [hunk 14]
```diff
@@ -0,0 +1,90 @@
...
+
+func (p *IBMProvider) GetName() string {
+	return "ibm"
+}
+
+func (p *IBMProvider) GetProviderData(arg ...string) map[string]interface{} {
+	return map[string]interface{}{
+		"provider": map[string]interface{}{
+			"ibm": map[string]interface{}{
+				"version": providerwrapper.GetProviderVersion(p.GetName()),
```
Comment (sergeylanzman) [inline providers/ibm/ibm_provider.go:44]: not need anymore https://github.com/GoogleCloudPlatform/terraformer/blob/6794a293fcb931c538821a3a67db274fdde89343/terraformutils/terraformoutput/hcl.go#L37

File: providers/ibm/instance_groups.go [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+	}
+
+	for _, instanceGroup := range allrecs {
+		instanceGoupID := *instanceGroup.ID
+		g.Resources = append(g.Resources, g.loadInstanceGroup(instanceGoupID, *instanceGroup.Name))
+		managers := make([]string, 0)
+		for i := 0; i < len(instanceGroup.Managers); i++ {
+			managers = append(managers, string(*(instanceGroup.Managers[i].ID)))
+		}
+		go g.handleManagers(sess, instanceGoupID, managers)
```
Comment (sergeylanzman) [inline providers/ibm/instance_groups.go:141]: I don't see any wait for finish work in all goroutines.

File: providers/ibm/instance_groups.go [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+		}
+		g.Resources = append(g.Resources, g.loadInstanceGroupManger(instanceGroupID, instanceGroupManagerID, *instanceGroupManager.Name))
+
+		policies := make([]string, 0)
+
+		for i := 0; i < len(instanceGroupManager.Policies); i++ {
+			policies = append(policies, string(*(instanceGroupManager.Policies[i].ID)))
+		}
+
+		go g.handlePolicies(sess, instanceGroupID, instanceGroupManagerID, policies)
```
Comment (sergeylanzman) [inline providers/ibm/instance_groups.go:110]: I don't see any wait for finish work in all goroutines.

File: providers/ibm/instance_groups.go [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+		}
+		go g.handleManagers(sess, instanceGoupID, managers)
+	}
+	close(g.importDone)
+}
+
+func (g *InstanceGroupGenerator) InitResources() error {
+	apiKey := os.Getenv("IBMCLOUD_API_KEY")
+	if apiKey == "" {
+		log.Fatal("No API key set")
```
Comment (sergeylanzman) [inline providers/ibm/instance_groups.go:149]: return error

File: providers/ibm/kp.go [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+		BaseURL: envFallBack([]string{"IBMCLOUD_KP_API_ENDPOINT"}, kpurl),
+		APIKey:  os.Getenv("IC_API_KEY"),
+		Verbose: kp.VerboseFailOnly,
+	}
+
+	client, err := kp.New(options, kp.DefaultTransport())
+	if err != nil {
+		return err
+	}
+	client.Config.InstanceID = "e63cc5b3-2594-4e13-b1b1-e8263095ff4b"
```
Comment (sergeylanzman) [inline providers/ibm/kp.go:56]: ??

File: providers/ibm/utils.go [hunk 19]
```diff
@@ -0,0 +1,111 @@
+package ibm
```
Comment (sergeylanzman) [inline providers/ibm/utils.go:1]: licence

File: providers/ibm/instance_groups.go [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+	}
+
+	g.importDone = make(chan bool)
+	g.fatalErrors = make(chan error)
+
+	go g.handleInstanceGroups(sess)
+
+	select {
+	case <-g.importDone:
+		log.Println("Done Importing")
```
Comment (sergeylanzman) [inline providers/ibm/instance_groups.go:170]: delete log

File: providers/ibm/instance_groups.go [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+	}
+
+	// Instantiate the service with an API key based IAM authenticator
+	sess, err := vpcv1.NewVpcV1(&vpcv1.VpcV1Options{
+		Authenticator: &core.IamAuthenticator{
+			ApiKey: apiKey,
+		},
+	})
+	if err != nil {
+		log.Fatal("Error creating VPC Service.")
```
Comment (sergeylanzman) [inline providers/ibm/instance_groups.go:159]: only return error

File: providers/ibm/instance_groups.go [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+
+	go g.handleInstanceGroups(sess)
+
+	select {
+	case <-g.importDone:
+		log.Println("Done Importing")
+		break
+	case err := <-g.fatalErrors:
+		close(g.fatalErrors)
+		log.Fatal("Error: ", err)
```
Comment (sergeylanzman) [inline providers/ibm/instance_groups.go:174]: only return error

**Comment by sergeylanzman** [comment]
@anilkumarnagaraj Good job! thanks!!
Few issues:
1. signed CLA
2. delete logs.println in resources
3. delete logs.fatal and return error. logs.Fatal exit program, if you want import many resources, one of them can close program
4. licence in files
5. goroutines issue - I don't find any line with wait for finished in all goroutines

**Comment by sergeylanzman** [comment]
@anilkumarnagaraj Can you signed CLA?

**Comment by anilkumarnagaraj** [comment]
@googlebot I signed it!

**Review by sergeylanzman — CHANGES_REQUESTED** [review]
3 issues:
1. https://github.com/GoogleCloudPlatform/terraformer/pull/663/files#diff-050bd2042edc200aeca4f15ef0cf63b438b1e1fbe9491f4b23ab2f8d4b59ef01R100
2. https://github.com/GoogleCloudPlatform/terraformer/pull/663/files#diff-8ec0974ea9817845502d6f7d5d9a1bb74ad7f8f23f4c96e54c2d7dad7b61e232R61
3. https://github.com/GoogleCloudPlatform/terraformer/pull/663/files#diff-63a9aae40d8899f9c4b20c1d7efadd157dd241250158a5298df3f307045e51c4R1

**Pull request merged**